libp2p/protocols/secure/noise.nim

```diff
@@ -0,0 +1,476 @@
+## Nim-LibP2P
+## Copyright (c) 2020 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import chronos
+import chronicles
+import nimcrypto/[utils, sysrand, sha2, hmac]
+import ../../connection
+import ../../peer
+import ../../protobuf/minprotobuf
+import secure,
+       ../../crypto/[crypto, chacha20poly1305, curve25519, hkdf]
+
+template unimplemented: untyped =
+  doAssert(false, "Not implemented")
+
+const
+  # https://godoc.org/github.com/libp2p/go-libp2p-noise#pkg-constants
+  NoiseCodec* = "/noise"
+  
+  PayloadString = "noise-libp2p-static-key:"
+
+  ProtocolXXName = "Noise_XX_25519_ChaChaPoly_SHA256"
+
+let
+  # Empty is a special value which indicates k has not yet been initialized.
+  EmptyKey: ChaChaPolyKey = [0.byte, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+  NonceMax = uint64.high - 1 # max is reserved
+
+type
+  KeyPair = object
+    privateKey: Curve25519Key
+    publicKey: Curve25519Key
+  
+  # https://noiseprotocol.org/noise.html#the-cipherstate-object
+  CipherState = object
+    k: ChaChaPolyKey
+    # noise spec says uint64
+    # go implementation has uint32 tho I noticed!!
+    n: uint64
+
+  # https://noiseprotocol.org/noise.html#the-symmetricstate-object
+  SymmetricState = object
+    cs: CipherState
+    ck: ChaChaPolyKey
+    h: MDigest[256]
+
+  # https://noiseprotocol.org/noise.html#the-handshakestate-object
+  HandshakeState = object
+    ss: SymmetricState
+    s: KeyPair
+    e: KeyPair
+    rs: Curve25519Key
+    re: Curve25519Key
+  
+  Noise* = ref object of Secure
+    localPrivateKey: PrivateKey
+    localPublicKey: PublicKey
+    noisePrivateKey: Curve25519Key
+    noisePublicKey: Curve25519Key
+
+  NoiseConnection* = ref object of Connection
+    insecure: Connection
+    readCs: CipherState
+    writeCs: CipherState
+
+  NoiseHandshakeError* = object of CatchableError
+
+# Utility
+
+proc genKeyPair: KeyPair =
+  result.privateKey = Curve25519Key.random()
+  result.publicKey = result.privateKey.public()
+    
+# todo this should be exposed maybe or replaced with a more public version
+proc getBytes*(key: string): seq[byte] =
+  result.setLen(key.len)
+  copyMem(addr result[0], unsafeaddr key[0], key.len)
+
+proc hashProtocol(name: string): MDigest[256] =
+  #[
```

Comment:
Not a biggie but if we can stick with one commenting style that would be great. We use single line comments like this:

```nim
proc myProc():
  ## this
  ## is my
  ## proc
```

Reply:
Fixed